larq/models.py

```diff
@@ -47,91 +67,316 @@ def _get_output_shape(layer):
+                ):
+                    count += op.n
+            return count
+        else:
+            return escape
+
+    def input_precision(self, default="-"):
+        try:
+            return self._layer.input_quantizer.precision
+        except:
+            return default
+
+    @property
+    def output_shape(self):
+        try:
+            return tuple(dim if dim else -1 for dim in self._layer.output_shape)
+        except AttributeError:
+            return "multiple"
+        except RuntimeError:  # output_shape unknown in Eager mode.
+            return "?"
+
+    @property
+    def output_pixels(self):
+        """Number of pixels for a single feature map (1 for fully connected layers)."""
+        if len(self.output_shape) == 4:
+            return int(np.prod(self.output_shape[1:3]))
+        elif len(self.output_shape) == 2:
+            return 1
+        else:
+            raise NotImplementedError()
+
+    @property
+    def unique_param_bidtwidths(self):
+        return sorted(set([p.bitwidth for p in self.parameter_profiles]))
+
+    @property
+    def unique_op_precisions(self):
+        return sorted(set([op.precision for op in self.op_profiles]))
+
+    def generate_table_row(self, table_config):
+        row = [self._layer.name, self.input_precision(), self.output_shape]
+        for i in table_config["param_bidtwidths"]:
+            n = self.parameter_count(i)
+            n = _format_table_entry(n, table_config["param_units"])
+            row.append(n)
+        row.append(_format_table_entry(self.memory, table_config["memory_units"]))
+        for i in table_config["mac_precisions"]:
+            n = self.op_count("mac", i)
+            n = _format_table_entry(n, table_config["mac_units"])
+            row.append(n)
+
+        return row
+
+    def _quantized_weights(self):
+        try:
+            return self._layer.quantized_latent_weights
+        except:
+            return []
+
+    def _get_bitwidth(self, weight):
+        try:
+            quantizer = self._layer.quantizers[self._quantized_weights().index(weight)]
+            return quantizer.precision
+        except:
+            return 32
+
+
+class ModelProfile:
+    def __init__(self, model):
+        self.layer_profiles = [LayerProfile(l) for l in model.layers]
+
+    @property
+    def memory(self):
+        return sum(l.memory for l in self.layer_profiles)
+
+    @property
+    def fp_equivalent_memory(self):
+        return sum(l.fp_equivalent_memory for l in self.layer_profiles)
+
+    def parameter_count(self, bitwidth=None, trainable=None):
+        return sum(l.parameter_count(bitwidth, trainable) for l in self.layer_profiles)
+
+    def op_count(self, op_type=None, bitwidth=None):
+        return sum(l.op_count(op_type, bitwidth, 0) for l in self.layer_profiles)
+
+    @property
+    def unique_param_bidtwidths(self):
+        return sorted(
+            set(_flatten([l.unique_param_bidtwidths for l in self.layer_profiles]))
```

Comment:
Does your `_flatten` also accept a generator expression?  If so I'd remove the `[` and `]`.